Add Fishing action and fish with master.

Celandriel opened the pull request:
With @notOrrytrout I started working on enabling bot fishing with master, but also on their own. 
The first commit didnt crash, showing that it was possible to have a bot cast when master does. Currently it compiles but crashes when you try to fish with a bot in the group, whether the bot has fishing or not. It makes me think that the check in FishingValues is broken somehow, but I cant figure out how.

DustinHendrickson commented on src/strategy/values/FishValues.cpp:
```diff
@@ -0,0 +1,103 @@
+            return false;
+        }
+    }
+    if (pole->GetSlot() != EQUIPMENT_SLOT_MAINHAND)
```
This path could still reference a nulptr for pole here , make sure its got a pole check before it.

if (!pole)
{
    return false;
}

notOrrytrout commented:
Im happy you took over. 🙂

notOrrytrout commented:
The Dustin changes seem promising too.

Celandriel commented:
So quite a few changes. Now the bots will fish with the master (Fishing on their own tbd) and the logic for detecting a bobber seems to be ok. 

Its not looting on pushing the worldpacket so that needs to be worked on, but im at a loss on how to do that right now.

notOrrytrout commented:
It would be cool to merge this. I know the bots can’t loot the bobber but at least they fish. Then make a separate issue that they don’t loot. What do you all think.

Celandriel commented:
Not just yet. I need to fix the windows compile issue, and I don't like that the action is triggered at every spell cast.

noisiver commented:
I think it makes sense to wait until the looting part is handled. I'm thinking of all the users who will complain about the bots not looting what they fish for as they don't know any better, not that I can fully disagree with them.

notOrrytrout commented:
Wow. You are good.  😎

Celandriel commented:
So I reworked quite a bit how this works. At this time I limited this to be triggered only when a master is fishing. Its not tested with rndbots (and either  way they dont have professions currently.)

The way this works is
The master casts fishing.
If the bot knows how to fish it will activate the fishing with master strategy. 
This checks if the bot is within 5 yards of water, if they are then they face the water. If not then they move within 10 yards of the water (set by the config)
Then it checks if the bot has a pole, and tried to equip one. (It will give an rndbot a fishing pole)
Then it casts the fishing spell, triggering the usebobber strategy. 
This will check the bobber state and use it when ready. 

If no bobber the bot owns is present, or if it loots the bobber, it turns off the usebobber strategy. 

The masterfishing strategy is turned off only when the master is further than 10 yards away from the bot. (If anyone has suggestions for additional end points feel free to suggest.)

hermensbas commented on src/strategy/values/FishValues.cpp:
```diff
@@ -0,0 +1,68 @@
+bool CanOpenBobberValue::Calculate()
+{
+  if (!bot || !botAI)
+    return false;
```
Can you add {} around the return false, having two return statements like that makes is very sensitive for error.

hermensbas commented:
Did anyone test the feature?

Celandriel commented:
So I made some under the hood changes, but have not been able to test in game since making them. Ill leave this as draft until I or someone else has time to test.

Celandriel commented:
So overall I made some fairly extensive changes since this was last looked at. 
Implemented packet changes.
Added a new helper functions to check for isfishingpole and if a specific spot has water using the map->GetLiquidData function which is more appropriate.
Changed the logic surrounding selecting the fishing spot, first it looks for any fishing nodes, then the master direction, and if that fails a search. Additionally, the search should try to make the bot be as perpendicular to shoreline as possible. 
Adjusted search ranges to match fishing spell of 10-20 yd range. 

Created equipFishingPoleAction.
Added facing water check to FishingAction
Add delay between casting spell and starting to check for bobber. 

Add remove bobber strategy, and set on a 20s timer. 

Adjusted priorities for fishing and bobber triggers down.
Move AI check for the bobber being ready to use to FishingValue
 

The bobber is now on a 20 second time

ScoobyPwnsOnU commented:
Tested on Horde.

-  Level 20 5-man mixed race group in Orgimmar Valley of Honor seems to **work correctly**.
-  They have a few **issues** correctly pointing themselves towards water as they will fix themselves if they're completely away from the water, but if they could make 1/10 casts then they won't better adjust themselves for a more consistent success rate. Same **issue** when trying to fish off the dock.
- Another **minor issue** is that they don't want to move themselves to the water to fish unless they're RIGHT ON IT. Even taking a few steps seems to be too far for them to attempt while under masterfishing. Screenshot below shows how far they were unwilling to move as I was standing about around the red circle and enabled master fishing, they refused to go over to the water on their own.
![image](https://github.com/user-attachments/assets/f980c89f-6c9b-43de-a2b2-41f44a55f468)
- Boosted my team up to level 80 and teleported to Borean Tundra where I ran into my first **bigger issue**. Bots think they require a certain fishing level to fish in the area, which is accurate to Vanilla, but not to the updated way fishing works in Wrath. The correct function would be for them to be able to fish anywhere as long as they have the fishing skill of any level, but with a low level of fishing they get worse loot results.
- I then tested attempting to fish as a selfbot, and it still **works somewhat correctly**. Triggering masterfishing on yourself will cause you to auto fish, though manually fishing **does not** seem to cause you to pick up the strategy, though it will cause your bots to still.
- I ran into my first **major issue** maybe a few hours into testing when I decided to test some alternative areas between the bottom and top of the difficulty fishing areas and went to Ashenvale first. Bots suddenly would not automatically take masterfishing upon your attempt to fish. If you manually set nc +masterfishing then they would fish as they were supposed to, but would not pick up the strategy automatically anymore. Attempts to fix this included a reset botAI, dropping and recreating the team multiple times, logging the bots off and on, going back to Valley of Honor where it previously worked, relogging the game, clearing the client cache, logging in with a different client, logging in with a different computer, rebooting the server, completely removing and deleting the pr from the server, recompiling, and then adding the pr back and recompiling again, creating a new character to see if it was impacting it too, trying with and without gm on, setting levels back down to 20 to see if it was something with level 80, and none of those things resolved the issue. I did resolve the issue by deleting my entire character database and recreating it, but I was unable to replicate the bug afterwards by retracing my steps. What they would output during debug when I fished during the bug is in the screenshot below.
![image](https://github.com/user-attachments/assets/a261197a-80e9-4c9e-9b45-4cef7ba1a0ac)

ScoobyPwnsOnU commented:
I believe I have narrowed down the source of the major bug of the bots not taking masterfishing when the master fishes.

It seems that bots specifically stop taking masterfishing when the master is above apprentice rank. I tested on two different accounts, setting my fishing level to 75 it worked, as soon as I bought the upgrade to journeyman fisher they stopped working.

Celandriel commented:
Ive made some fairly heavy changes to how this worked. 

1. Merge detection of master casting the fishing spell in the SeeSpellsAction (no need for duplicate packet handling)
2. Reduce priorities for actions down to reasonable numbers. (https://github.com/liyunfan1223/mod-playerbots/blob/a19604024eaa424fdddaac9f3ae2a60b67047d05/src/strategy/Strategy.h#L43)
3. Set removal of fishing strategy only as a distance check from water. Strategy gets auto removed if beyond 40.
4. Add support for Fishing Holes and updated code naming convention. 
6. Changed search functions for land and water. it now uses a circular pattern with 2pi/16 
7. Added linear search for finding water/land between two points. 
8. Fix bug where master fishing level was limited to apprentice level.
Ready to test.

Celandriel commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,361 @@
+bool EquipFishingPoleAction::isUseful()
+{
+    if (!bot)
+        return false;
```
I think it was an early recommendation to check against weird issues...

Celandriel commented on src/strategy/values/FishValues.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (SkillFishing == 0)
+        return false;
+
+    return true;
```
Expanded conditions, left as is.

Celandriel commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,361 @@
+        }
+        else
+        {
+            botAI->TellError("I don't have a fishing pole");
```
Do we have a standard approach for localized messages? I couldnt tell from the code.

brighton-chi replied:
Look at my Karazhan PR for yells relating to the Netherspite beam blocking actions.

Gromleq made a new method for this purpose now.

ScoobyPwnsOnU commented:
Updates seem to have fixed the issue with using fishing above apprentice, and also the hard crash anytime bots received masterfishing is also fixed, but now the server is hanging up sometime while i was afk for a few minutes. Not sure what caused it as I wasn't near water or fishing at that point in time.

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -52,7 +52,7 @@ float hasFishableWaterOrLand(float x, float y, float z,  Map* map, uint32 phaseM
     LiquidData const& liq = map->GetLiquidData(phaseMask, x, y, z+HEIGHT_ABOVE_WATER_TOLERANCE, DEFAULT_COLLISION_HEIGHT, MAP_ALL_LIQUIDS);
     float ground = map->GetHeight(phaseMask, x, y, z + HEIGHT_SEARCH_BUFFER, true);
-
+;
```
empty line with ; ?

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -142,19 +155,23 @@ WorldPosition findLandRadialFromPosition (Player* bot, WorldPosition targetPos,
             {
+                if (master && botAI->HasStrategy("follow", BOT_STATE_NON_COMBAT))
+                {
+                    if (master->GetDistance(checkX, checkY, groundZ) > fishingSearchWindow - 0.5)
```
You can combine this 2 if's

kadeshar commented on src/strategy/values/FishValues.h:
```diff
@@ -0,0 +1,48 @@
+        value = val;
+        _setTime = getMSTime();
+    }
+    uint32 lastUpdated() const {return - _setTime;}
```
This minus is correct?

kadeshar commented on src/TravelMgr.cpp:
```diff
@@ -218,6 +218,11 @@ bool WorldPosition::isUnderWater()
 
+bool WorldPosition::IsValid()
+{
+    return GetPositionX() != 0 && GetPositionY() != 0;
```
There is no possibility that Position (0, 0) is not valid? Checking mapId not help?

kadeshar commented on src/PlayerbotAIConfig.h:
```diff
@@ -86,7 +86,7 @@ class PlayerbotAIConfig
     float sightDistance, spellDistance, reactDistance, grindDistance, lootDistance, shootDistance, fleeDistance,
         tooCloseDistance, meleeDistance, followDistance, whisperDistance, contactDistance, aoeRadius, rpgDistance,
-        targetPosRecalcDistance, farDistance, healDistance, aggroDistance;
+        targetPosRecalcDistance, farDistance, healDistance, aggroDistance, fishingDistanceFromMaster, fishingDistance;
```
I dont see aggroDistance in src/PlayerbotAIConfig.cpp and usage in code

kadeshar commented on src/strategy/actions/ActionContext.h:
```diff
@@ -195,7 +195,7 @@ class ActionContext : public NamedObjectContext<Action>
         creators["go fishing"] = &ActionContext::go_fishing;
         creators["use fishing bobber"] = &ActionContext::use_fishing_bobber;
-        creators["end fishing"] = &ActionContext::end_fishing;
+        creators["end master fishing"] = &ActionContext::end_fishing;
```
end_master_fishing?

Celandriel commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,504 @@
+        return false;
+    FishingSpotValue* fishingSpotValueObject =  (FishingSpotValue*)context->GetValue<WorldPosition>("fishing spot"); 
+    WorldPosition pos = fishingSpotValueObject->Get();
+    return pos.IsValid() || fishingSpotValueObject->IsStale(180000) || pos != bot->GetPosition();
```
The action is useful under three cases, the position is not valid, the position is stale, or the bot is not at the stored position.

noisiver commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,501 @@
+bool MoveNearWaterAction::Execute(Event event)
+{
+    WorldPosition landSpot = AI_VALUE(WorldPosition, "fishing spot");
+    if(landSpot.IsValid())
```
Another instance where brackets are used for a single line if.

noisiver commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,501 @@
+bool EquipFishingPoleAction::isUseful()
+{
+    Item* mainHand = bot->GetItemByPos(INVENTORY_SLOT_BAG_0, EQUIPMENT_SLOT_MAINHAND);
+    if (IsFishingPole(mainHand))
```
Another instance where brackets are used for a single line if.

noisiver commented:
Aside from the few reviews I submitted - which are just visual - I don't see a problem merging this. Perhaps it makes sense to resolve the places I mentioned before I approve it?

Celandriel commented:
Ill give this another pass to see if can button up the stormwind issue.

noisiver commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,497 @@
+bool MoveNearWaterAction::Execute(Event event)
+{
+    WorldPosition landSpot = AI_VALUE(WorldPosition, "fishing spot");
+    if(landSpot.IsValid())
```
No space after the word if

kadeshar commented on src/strategy/generic/WorldPacketHandlerStrategy.cpp:
```diff
@@ -59,7 +59,8 @@ void WorldPacketHandlerStrategy::InitTriggers(std::vector<TriggerNode*>& trigger
     // relevance), nullptr)));
     triggers.push_back(new TriggerNode("group list", NextAction::array(0, new NextAction("reset botAI", relevance), nullptr)));
-    triggers.push_back(new TriggerNode("see spell", NextAction::array(0, new NextAction("see spell", relevance), nullptr)));
+    triggers.push_back(new TriggerNode("see spell", NextAction::array(0, new NextAction("see spell", relevance), new NextAction("fish with master", relevance), nullptr)));
```
I dont see action "fish with master"

Celandriel commented:
I fixed the WorldPosition check. I had misread the initialization previously. Works now.

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,492 @@
+{
+    if (!AI_VALUE(bool, "can fish"))
+        return false;
+    FishingSpotValue* fishingSpotValueObject =  (FishingSpotValue*)context->GetValue<WorldPosition>("fishing spot");
```
Delete extra space

kadeshar replied:
Delete extra space

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,492 @@
+        float distance = bot->GetExactDist2d(&fishingHole);
+        bool hasLOS = bot->IsWithinLOS(fishingHole.GetPositionX(), fishingHole.GetPositionY(), fishingHole.GetPositionZ());
+        // Water spot is in range, and we have LOS to it. Set bot position to fishing spot and do not move
+        if (distance  >= MIN_DISTANCE_TO_WATER &&
```
Delete extra space

kadeshar replied:
Delete extra space

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,492 @@
+        bool hasLOS = bot->IsWithinLOS(fishingHole.GetPositionX(), fishingHole.GetPositionY(), fishingHole.GetPositionZ());
+        // Water spot is in range, and we have LOS to it. Set bot position to fishing spot and do not move
+        if (distance  >= MIN_DISTANCE_TO_WATER &&
+            distance <=  MAX_DISTANCE_TO_WATER && hasLOS)
```
Delete extra space

kadeshar commented on src/strategy/actions/FishingAction.h:
```diff
@@ -0,0 +1,72 @@
+    bool isUseful() override;
+};
+
+class EndMasterFishing : public Action
```
Sometimes you use word Action as method suffix sometimes no. I prefer using.

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,492 @@
+{
+    if (!AI_VALUE(bool, "can fish"))
+        return false;
+    FishingSpotValue* fishingSpotValueObject =  (FishingSpotValue*)context->GetValue<WorldPosition>("fishing spot");
```
Delete extra space

kadeshar replied:
Delete extra space

kadeshar commented on src/strategy/actions/ActionContext.h:
```diff
@@ -380,6 +386,11 @@ class ActionContext : public NamedObjectContext<Action>
+    static Action* go_fishing(PlayerbotAI* botAI) { return new FishingAction(botAI);}
+    static Action* use_fishing_bobber(PlayerbotAI* botAI) { return new UseBobber(botAI);}
+    static Action* end_master_fishing(PlayerbotAI* botAI) { return new EndMasterFishing(botAI); }
+    static Action* remove_bobber_strategy(PlayerbotAI* botAI) { return new RemoveBobberStrategyAction(botAI); }
```
Sometimes you use Action as method suffix sometimes not. Be consistent. Personally prefer using suffix.

kadeshar commented on src/strategy/actions/FishingAction.cpp:
```diff
@@ -0,0 +1,492 @@
+
+bool EndMasterFishing::isUseful()
+{
+    FishingSpotValue* fishingSpotValueObject =  (FishingSpotValue*)context->GetValue<WorldPosition>("fishing spot");
```
Delete extra space

notOrrytrout commented:
Need a review to merge before Xmas

brighton-chi commented:
It is 4:45 pm AoE. There's still time.

NoxMax commented:
The SQL file still needs the minor update for ID and date.

kadeshar commented on data/sql/playerbots/updates/2025_10_13_ai_playerbot fishing text.sql:
```diff
@@ -0,0 +1,15 @@
+DELETE FROM ai_playerbot_texts_chance WHERE name IN ('no_fishing_pole_error');
+
+INSERT INTO ai_playerbot_texts (id, name, text, say_type, reply_type, text_loc1, text_loc2, text_loc3, text_loc4, text_loc5, text_loc6, text_loc7, text_loc8) VALUES
+(1717, 'no_fishing_pole_error',   "I don't have a Fishing Pole",   0, 0, 
```
Next free id is 1736. Please change that
